src/charm.py

```diff
@@ -30,6 +31,60 @@ def __init__(self, *args: typing.Any) -> None:
         super().__init__(*args)
         self.framework.observe(self.on.config_changed, self.reconciliation)
 
+    def _application_config(self) -> dict | None:
+        """Decode the value of the charm configuration application-config.
+
+        Returns:
+            The value of the charm configuration application-config.
+
+        Raises:
+            ReconciliationError: when application-config is invalid.
```

Comment:
Would something like an `InputError` or something like that be better here? Reconciliation error sounds like something went wrong when the charm tried to implement changes rather than user input being incorrect?

Reply:
`ReconciliationError` is the mechanism for reconciliation subroutines to notify the main reconciliation process that something has gone wrong and the reconciliation need to terminate early and gracefully. 
The information about the error is saved in the `new_status` attribute of `ReconciliationError` and it will be used by the main reconciliation process to update the charm status.